### PR TITLE
GEODE-4424: Refactor PoolManager::createFactory to return value

### PR DESCRIPTION
**Pull request by mmartell**
Testing: Unit/integration tests passed on Mac and Windows

### CODE REVIEW

File: clicache/src/Cache.cpp [hunk 1]
```diff
@@ -339,7 +339,8 @@ namespace Apache
 
         PoolFactory^ Cache::GetPoolFactory()
         {
-          return PoolFactory::Create(m_nativeptr->get_shared_ptr()->getPoolManager().createFactory());
+          return PoolFactory::Create(std::make_shared<native::PoolFactory>(
```
Comment (pivotal-jbarrett) [inline clicache/src/Cache.cpp]: The .NET factory does not need to keep a shared pointer to the C++ factory. It is a 1:1 relationship now so it should keep a unique_ptr. See other factories that have gone through this change for examples.

File: clicache/src/PoolManager.cpp [hunk 2]
```diff
@@ -37,7 +37,8 @@ namespace Apache
 
       PoolFactory^ PoolManager::CreateFactory()
       {
-        return PoolFactory::Create(m_nativeref.createFactory());
+        return PoolFactory::Create(std::make_shared<native::PoolFactory>(
```
Comment (pivotal-jbarrett) [inline clicache/src/PoolManager.cpp]: Same, use unique_ptr.

File: cppcache/include/geode/PoolFactory.hpp [hunk 3]
```diff
@@ -533,9 +533,10 @@ class _GEODE_EXPORT PoolFactory {
 
   ~PoolFactory();
 
+  PoolFactory(const PoolFactory&);
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/PoolFactory.hpp]: If the copy is trivial, as it should be just default, inline and include I header.

File: cppcache/include/geode/PoolFactory.hpp [hunk 4]
```diff
@@ -544,6 +545,7 @@ class _GEODE_EXPORT PoolFactory {
 
   friend class Cache;
   friend class PoolManager;
+  friend class PoolManagerImpl;
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/PoolFactory.hpp]: Why this friend?

File: cppcache/integration-test/CacheHelper.cpp [hunk 5]
```diff
@@ -134,7 +134,7 @@ CacheHelper::CacheHelper(const std::shared_ptr<Properties>& configPtr,
   auto cacheFactory = CacheFactory(pp);
   cachePtr = std::make_shared<Cache>(cacheFactory.create());
 
-  auto poolFactory = cachePtr->getPoolManager().createFactory();
+  auto poolFactory = std::make_shared<PoolFactory>(cachePtr->getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/CacheHelper.cpp]: Is the pool factory assigned to a member variable later? If not then we don’t need the shared_ptr.

File: cppcache/integration-test/testPoolFactory.cpp [hunk 6]
```diff
@@ -29,7 +29,7 @@ BEGIN_TEST(POOLFACTORY)
 {
   auto cacheFactory = CacheFactory();
   auto cache = cacheFactory.create();
-  auto poolFactory = cache.getPoolManager().createFactory();
+  auto poolFactory = std::make_shared<PoolFactory>(cache.getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/testPoolFactory.cpp]: If this is only a local variable then you don’t need to make it shared_ptr.

File: cppcache/src/PoolFactory.cpp [hunk 7]
```diff
@@ -69,6 +69,12 @@ PoolFactory::PoolFactory(const Cache& cache)
       m_addedServerOrLocator(false),
       m_cache(cache) {}
 
+PoolFactory::PoolFactory(const PoolFactory& factory)
```
Comment (pivotal-jbarrett) [inline cppcache/src/PoolFactory.cpp]: This looks like default impl, use that instead.

File: clicache/src/Cache.cpp (hunk 1 above)
Comment (dgkimura) [reply clicache/src/Cache.cpp]: Thanks. Agreed that unique_ptr is better here.

File: cppcache/include/geode/PoolFactory.hpp (hunk 3 above)
Comment (dgkimura) [reply cppcache/include/geode/PoolFactory.hpp]: Yup. Done.

File: cppcache/include/geode/PoolFactory.hpp (hunk 4 above)
Comment (dgkimura) [reply cppcache/include/geode/PoolFactory.hpp]: `PoolManagerImpl::createFactory` calls into `PoolFactory` constructor and `PoolFactory` constructor is private.  I've wondered is if these constructors even need to be private anymore?  I assumed it used to be to protect the original singleton design.  Perhaps this should be a separate ticket to investigate.

File: cppcache/integration-test/CacheHelper.cpp (hunk 5 above)
Comment (dgkimura) [reply cppcache/integration-test/CacheHelper.cpp]: It is passed around later on through method calls. We had kept it as shared_ptr to minimize our code changes.  But your requested change isn't that bad so it has been updated in next revision.

File: cppcache/include/geode/PoolFactory.hpp (hunk 4 above)
Comment (pivotal-jbarrett) [reply cppcache/include/geode/PoolFactory.hpp]: Ok. It is more likely private to require the use of the factory method. Just odd to expose internal classes in a public header like this. Should be ok for now.

File: cppcache/integration-test/testPoolFactory.cpp [hunk 8]
```diff
@@ -30,9 +30,8 @@ BEGIN_TEST(POOLFACTORY)
   auto cacheFactory = CacheFactory();
   auto cache = cacheFactory.create();
   auto poolFactory = cache.getPoolManager().createFactory();
-  ASSERT(poolFactory != nullptr, "poolFactory was a nullptr");
-  auto& testPoolFactory = poolFactory->setSubscriptionEnabled(true);
-  bool test = poolFactory.get() == std::addressof<PoolFactory>(testPoolFactory);
+  auto& testPoolFactory = poolFactory.setSubscriptionEnabled(true);
+  bool test = &poolFactory == std::addressof<PoolFactory>(testPoolFactory);
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/testPoolFactory.cpp]: Do you think this test is really valuable anymore? Maybe PoolFactory should implement `operator==()` and we can compare them this way?

File: cppcache/integration-test/testThinClientCqDurable.cpp [hunk 9]
```diff
@@ -192,7 +192,7 @@ void RunDurableCqClient() {
   // Create a Geode Cache Programmatically.
   auto cacheFactory = CacheFactory(pp);
   auto cachePtr = std::make_shared<Cache>(cacheFactory.create());
-  auto poolFactory = cachePtr->getPoolManager().createFactory();
+  auto poolFactory = std::make_shared<PoolFactory>(cachePtr->getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/testThinClientCqDurable.cpp]: Is `shared_ptr` necessary?

File: cppcache/integration-test/testThinClientPoolAttrTest.cpp [hunk 10]
```diff
@@ -69,7 +69,7 @@ class putThread : public ACE_Task_Base {
 };
 
 void doAttrTestingAndCreatePool(const char* poolName) {
-  auto poolFacPtr = getHelper()->getCache()->getPoolManager().createFactory();
+  auto poolFacPtr = std::make_shared<PoolFactory>(getHelper()->getCache()->getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/testThinClientPoolAttrTest.cpp]: Is `shared_ptr` necessary?

File: cppcache/integration-test/testThinClientPoolAttrTest.cpp [hunk 11]
```diff
@@ -204,7 +204,7 @@ DUNIT_TASK(CLIENT1, StartC1)
    doAttrTestingAndCreatePool(poolName);
 
    // Do PoolCreation testing , create another pool with same name
-   auto poolFacPtr = getHelper()->getCache()->getPoolManager().createFactory();
+   auto poolFacPtr = std::make_shared<PoolFactory>(getHelper()->getCache()->getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/testThinClientPoolAttrTest.cpp]: Is `shared_ptr` necessary?

File: cppcache/integration-test/testThinClientSecurityMultiUserTest.cpp [hunk 12]
```diff
@@ -47,7 +47,7 @@ END_TASK_DEFINITION
 DUNIT_TASK_DEFINITION(CLIENT1, PerformSecureOperationsWithUserCredentials)
   {
     auto cache = CacheFactory().create();
-    auto poolFactory = cache.getPoolManager().createFactory();
+    auto poolFactory = std::make_shared<PoolFactory>(cache.getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/testThinClientSecurityMultiUserTest.cpp]: Is `shared_ptr` necessary?

File: cppcache/src/CacheXmlParser.cpp [hunk 13]
```diff
@@ -597,7 +597,7 @@ void CacheXmlParser::startPool(const xmlChar** atts) {
         "A pool cannot be created without a name";
     throw CacheXmlException(s.c_str());
   }
-  auto factory = m_cache->getPoolManager().createFactory();
+  auto factory = std::make_shared<PoolFactory>(m_cache->getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline cppcache/src/CacheXmlParser.cpp]: Is `shared_ptr` necessary?

File: tests/cpp/fwklib/FrameworkTest.cpp [hunk 14]
```diff
@@ -338,7 +338,7 @@ void FrameworkTest::localDestroyRegion(std::shared_ptr<Region>& region) {
 void FrameworkTest::parseEndPoints(int32_t ep, std::string label,
                                    bool isServer) {
   std::string poolName = "_Test_Pool";
-  auto pfPtr = m_cache->getPoolManager().createFactory();
+  auto pfPtr = std::make_shared<PoolFactory>(m_cache->getPoolManager().createFactory());
```
Comment (pivotal-jbarrett) [inline tests/cpp/fwklib/FrameworkTest.cpp]: Is `shared_ptr` necessary?

File: tests/cpp/fwklib/FwkObjects.cpp [hunk 15]
```diff
@@ -573,7 +573,7 @@ FwkPool::FwkPool(const DOMNode* node) : m_locators(false), m_servers(false) {
   // Init Factory
   m_cache = std::make_shared<Cache>(CacheFactory().create());
   m_poolManager = &m_cache->getPoolManager();
-  m_poolFactory = m_poolManager->createFactory();
+  m_poolFactory = std::make_shared<PoolFactory>(m_poolManager->createFactory());
```
Comment (pivotal-jbarrett) [inline tests/cpp/fwklib/FwkObjects.cpp]: Is `shared_ptr` necessary?

File: cppcache/integration-test/testPoolFactory.cpp (hunk 8 above)
Comment (dgkimura) [reply cppcache/integration-test/testPoolFactory.cpp]: I had wondered this as well.  My interpretation is that this test is trying to assert that the fluent interface is returning the exact same object which is probably not how one would usually expect `operator==()` to work.  I am fine deleting this test if you agree that it has minimal value.

File: cppcache/src/CacheXmlParser.cpp (hunk 13 above)
Comment (dgkimura) [reply cppcache/src/CacheXmlParser.cpp]: Yes.  In this case `shared_ptr` is providing abstraction to push objects onto a `std::stack<std::shared_ptr<void>>` _stack.

File: cppcache/integration-test/testThinClientCqDurable.cpp (hunk 9 above)
Comment (dgkimura) [reply cppcache/integration-test/testThinClientCqDurable.cpp]: Not necessary. Just minimal change to get the code working.  Probably best to keep test code up-to-date so I'll see about making suggested changes to remove `shared_ptr` in this context in integration-tests.

File: cppcache/integration-test/testPoolFactory.cpp (hunk 8 above)
Comment (pivotal-jbarrett) [reply cppcache/integration-test/testPoolFactory.cpp]: Oh.. I misread what it was doing. Yes leave it.
👍

**Pull request merged**